nimsuggest/tests/terror_handling.nim

```diff
@@ -0,0 +1,18 @@
+# Test ill formed AST will not causes crash and subsequent errors
+
+import std/macros
+
+macro t1(): untyped =
+  result = newNimNode(nnkForStmt)
+  result.add(ident("i")) 
+  result.add newNimNode(nnkCall)
+  result.add nnkDiscardStmt.newTree(newEmptyNode())
```

Comment:
no, the purpose here is ensure for statement errors not be reported.

Reply:
The issue you're fixing is not at all related to `for` statements.

You didn't provide a reproducible case for the `IndexDefect` errors you encountered within `semFor`, but I'm very certain that this PR doesn't fix them. If the error would be caused by a macro-generated zero-length `nkCall` node, then the compiler would have already crashed in `semExpr` when analysing the `nkCall` node.

Don't get me wrong, I still think that making sure that `checkSonsLen` always at least raises an exception is a very important fix.

Reply:
I still think it's needed, `newNimNode` call doesn't check the length.

Reply:
Yes, `checkSonsLen` and `checkSonsMinLen` should always at least raise an error.

Reply:
> the test case covered before this PR, so I want ensure it will not happen again.

I don't understand. As you can see by what the `chk` result is reported for (the invalid `nkCall` node), the `nnkForStmt` is unnecessary.

If you mean that the test should use `nnkCall.newNimNode()` instead of `nnkCall.newTree()`, then yep, that's better.

Reply:
but before this PR, the test case will also cause `IndexDefect` in `semFor`, so my original attemption is not only the nnkCall error reported, but its subsequent errros in `semFor`

Reply:
> but before this PR, the test case will also cause `IndexDefect` in `semFor`

No, `semFor` doesn't crash with an `IndexDefect` prior to this PR, and yes, I've tested it. `nimsuggest` crashes in `semExpr`, when analysing the `nkCall` node, like I said above.

The original intention is not relevant for the final a test. The final test has to focus  on the actually fixed issue, and not be what the investigation into the issue started with. In other words, the test needs to be reduced as much as possible, and as it stands, your test is not. Please use the suggestion I provided -- I've also changed it to use `newNimNode`.

If `checkMinSonsLen` wouldn't raise then my suggested test would cause an `IndexDefect` crash, in the very same way that the current test does too.

Strictly speaking, each node kind would need to be tested here, but testing one is enough for now.